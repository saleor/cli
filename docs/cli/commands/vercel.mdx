# vercel

## Integrate with Vercel

### vercel

```sh
$ saleor vercel --help
```

Help output:

```
saleor vercel [command]

Integrate with Vercel

Commands:
  saleor vercel login  Add integration for Saleor CLI

Options:
      --json             Output the data as JSON  [boolean] [default: false]
      --short            Output data as text  [boolean] [default: false]
  -u, --instance, --url  Saleor instance API URL (must start with the protocol, i.e. https:// or http://)  [string]
  -V, --version          Show version number  [boolean]
  -h, --help             Show help  [boolean]
```

#### vercel login

```sh
$ saleor vercel login --help
```

Help output:

```
saleor vercel login

Add integration for Saleor CLI

Options:
      --json             Output the data as JSON  [boolean] [default: false]
      --short            Output data as text  [boolean] [default: false]
  -u, --instance, --url  Saleor instance API URL (must start with the protocol, i.e. https:// or http://)  [string]
  -V, --version          Show version number  [boolean]
  -h, --help             Show help  [boolean]
```