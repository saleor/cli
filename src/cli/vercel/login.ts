import chalk from 'chalk';
import Debug from 'debug';
import EventEmitter from 'events';
import got from 'got';
import { nanoid } from 'nanoid';
import { ServerApp } from 'retes';
import { Response } from 'retes/response';
import { GET } from 'retes/route';

import { Config, SaleorCLIPort } from '../../lib/config.js';
import { checkPort } from '../../lib/detectPort.js';
import { NoCommandBuilderSetup } from '../../lib/index.js';
import { delay, openURL, printlnSuccess, successPage } from '../../lib/util.js';

const RedirectURI = `http://localhost:${SaleorCLIPort}/vercel/callback`;

const debug = Debug('saleor-cli:vercel:login');

export const command = 'login';
export const desc = 'Add integration for Saleor CLI';

export const builder = NoCommandBuilderSetup;

export const handler = async () => {
  await checkPort(SaleorCLIPort);

  const generatedState = nanoid();
  const emitter = new EventEmitter();

  const { VercelClientID, VercelClientSecret } = await Config.get();

  // const spinner = ora('\nLogging in...').start();
  // await delay(1500);
  // spinner.text = '\nLogging in...\n';

  const QueryParams = new URLSearchParams({ state: generatedState });
  const url = `https://vercel.com/integrations/saleor-cli/new?${QueryParams}`;
  await openURL(url);

  const app = new ServerApp([
    GET('/vercel/callback', async ({ params }) => {
      const { state, code } = params;

      if (state !== generatedState) {
        return Response.BadRequest('Wrong state');
      }

      const Params = {
        client_id: VercelClientID,
        client_secret: VercelClientSecret,
        code,
        redirect_uri: RedirectURI,
      };

      try {
        const data: any = await got
          .post('https://api.vercel.com/v2/oauth/access_token', {
            form: Params,
          })
          .json();

        const { access_token: accessToken, team_id: teamId } = data;

        await Config.set('vercel_token', `Bearer ${accessToken}`);
        await Config.set('vercel_team_id', teamId);
        printlnSuccess(chalk.bold('success'));
      } catch (error: any) {
        console.log(error.message);
        console.log(
          chalk(
            'Tip: in some cases',
            chalk.green('saleor logout'),
            'followed by',
            chalk.green('saleor login'),
            'may help'
          )
        );

        emitter.emit('finish');

        return {
          body: successPage('Login failed!'),
          status: 200,
          type: 'text/html',
          headers: {
            'Content-Type': 'text/html; charset=utf-8',
          },
        };
      }

      // spinner.succeed(`You've successfully logged into Saleor Cloud!\n  Your access token has been safely stored, and you're ready to go`)
      emitter.emit('finish');

      return {
        body: successPage(
          'You\'ve successfully authenticated Vercel with the Saleor CLI!'
        ),
        status: 200,
        type: 'text/html',
        headers: {
          'Content-Type': 'text/html; charset=utf-8',
        },
      };
    }),
  ]);
  await app.start(SaleorCLIPort);

  emitter.on('finish', async () => {
    await delay(1500);
    await app.stop();
  });
};
