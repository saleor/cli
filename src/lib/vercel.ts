import chalk from 'chalk';
import fetch from 'node-fetch';
import ora from 'ora';

import { delay, SaleorAppError } from './util.js';

export interface Env {
  key: string;
  value: string;
  target?: string[];
  type?: string;
}

export interface Deployment {
  id: string;
  uid: string;
  url: string;
  readyState: string;
  alias: string;
  errorCode?: string;
  inspectorUrl: string;
}

// eslint-disable-next-line import/prefer-default-export
export class Vercel {
  _client: any;

  constructor(accessToken: string, teamId: string | null | undefined = null) {
    this._client = async (
      method = 'GET',
      path = '',
      body: unknown = undefined
    ) => {
      const dest = `https://api.vercel.com${path}${
        teamId ? `?teamId=${teamId}` : ''
      }`;

      const r = await fetch(dest, {
        method,
        headers: {
          'Content-Type': 'application/json',
          Authorization: accessToken,
        },
        body: JSON.stringify(body),
      });

      // FIXME rethink the approach
      const result = (await r.json()) as any;

      if (!r.status.toString().startsWith('2')) {
        const { error } = result;
        if (error) {
          throw new Error(error.message);
        }
      }

      return result;
    };
  }

  async addEnvironmentVariables(projectId: string, envs: Env[]) {
    return this._client(
      'POST',
      `/v9/projects/${projectId}/env`,
      envs.map((env) => ({
        type: 'plain',
        target: ['production', 'development', 'preview'],
        ...env,
      }))
    );
  }

  async removeEnvironmentVariables(projectId: string, envs: Env[]) {
    const { envs: existingEnvs = [] }: any = await this._client(
      'GET',
      `/v9/projects/${projectId}/env`
    );

    for (const env of envs) {
      const byName = (key: string) => (el: any) => el.key === key;
      const envToRemove = existingEnvs.filter(byName(env.key)).shift();

      if (envToRemove) {
        await this._client(
          'DELETE',
          `/v9/projects/${projectId}/env/${envToRemove.id}`
        );
      }
    }
  }

  async setEnvironmentVariables(projectId: string, envs: Env[]) {
    await this.removeEnvironmentVariables(projectId, envs);
    await this.addEnvironmentVariables(projectId, envs);
  }

  async createProject(
    name: string,
    envs: Env[],
    owner: string,
    repoName: string,
    buildCommand: null | string = null,
    rootDirectory: null | string = null,
    provider = 'github'
  ) {
    const {
      framework: { slug },
    } = await this.detectFramework(owner, repoName, provider);

    return this._client('POST', '/v9/projects', {
      name,
      environmentVariables: envs,
      gitRepository: {
        type: provider,
        repo: `${owner}/${repoName}`,
        // sourceless: true, // FIXME
      },
      framework: slug || 'nextjs',
      buildCommand,
      rootDirectory,
    });
  }

  async deploy(name: string, provider: string, repoId: string) {
    const response = await this._client('POST', '/v13/deployments', {
      gitSource: {
        type: provider,
        ref: 'main',
        repoId,
      },
      name,
      target: 'production',
      source: 'import',
    });

    return response;
  }

  async getDeployments(projectId: string) {
    return this._client('GET', `/v6/deployments?projectId=${projectId}`) as {
      deployments: Deployment[];
    };
  }

  async getDeployment(deploymentId: string) {
    return this._client(
      'GET',
      `/v13/deployments/${deploymentId}`
    ) as Deployment;
  }

  async verifyDeployment(
    name: string,
    deploymentId: string,
    msg = 'Deployment of'
  ) {
    const spinner = ora(`${msg} ${chalk.cyan(name)} in progress...`).start();

    let readyState;
    do {
      const deployment = await this.getDeployment(deploymentId);
      readyState = deployment.readyState;

      if (hasDeploymentFailed(readyState)) {
        const { errorCode, inspectorUrl } = deployment;
        console.log(
          `\nVercel deployment status : ${readyState} - ${errorCode}`
        );
        console.log('Verify deployment using the following link:');
        console.log(inspectorUrl);

        throw new SaleorAppError();
      }

      await delay(5000);
    } while (hasDeploymentSucceeded(readyState));

    spinner.succeed(`Deployed ${chalk.cyan(name)}\n`);
  }

  async detectFramework(owner: string, repoName: string, provider: string) {
    const URL = `https://${provider}.com/${owner}/${repoName}`;

    return this._client('GET', `/v1/integrations/detect-framework?url=${URL}`);
  }

  async getProject(name: string) {
    return this._client('GET', `/v9/projects/${name}`);
  }

  async getProjectDomains(projectID: string) {
    return this._client('GET', `/v9/projects/${projectID}/domains`);
  }

  async getProjectDomain(projectID: string) {
    const { domains } = await this.getProjectDomains(projectID);

    return domains[0];
  }
}

const hasDeploymentSucceeded = (readyState: string) => readyState !== 'READY';
const hasDeploymentFailed = (readyState: string) =>
  ['ERROR', 'CANCELED'].includes(readyState);
