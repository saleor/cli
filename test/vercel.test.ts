import { rest } from 'msw';
import { setupServer } from 'msw/node';
import {
  afterAll,
  afterEach,
  assert,
  beforeAll,
  describe,
  expect,
  it,
} from 'vitest';

import { Vercel } from '../src/lib/vercel.js';

const VercelURL = 'https://api.vercel.com';

const myProject1 = {
  name: 'my-project-1',
  slug: 'my-project-1',
};

const handlers = [
  rest.get('https://rest.example/path/to/posts', (req, res, ctx) =>
    res(ctx.status(200), ctx.json(myProject1)),
  ),
  rest.get(`${VercelURL}/v9/projects/my-project-1`, (req, res, ctx) =>
    res(ctx.status(200), ctx.json(myProject1)),
  ),
  rest.get(`${VercelURL}/v13/deployments/xxx`, (req, res, ctx) =>
    res(ctx.status(200), ctx.json([])),
  ),
];

const server = setupServer(...handlers);
server.printHandlers();

const vercel = new Vercel('asdf');

beforeAll(() => server.listen({ onUnhandledRequest: 'error' }));
afterAll(() => server.close());
afterEach(() => server.resetHandlers());

describe('Vercel API', () => {
  it('returns `id` for single project', async () => {
    const r = await vercel.getProject('my-project-1');

    expect(r).toEqual({ name: 'my-project-1', slug: 'my-project-1' });
  });

  it('foo', () => {
    assert.equal(Math.sqrt(4), 2);
  });

  it('bar', () => {
    expect(1 + 1).eq(2);
  });

  it('snapshot', () => {
    expect({ foo: 'bar' }).toMatchSnapshot();
  });
});
