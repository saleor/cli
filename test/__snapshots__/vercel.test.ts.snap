// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vercel API > snapshot 1`] = `
{
  "foo": "bar",
}
`;
